### ISSUE TITLE
Avoid keeping two copies of gradients (param.grad and buckets) in DDP

### ISSUE THREAD

**Issue by mrshenli** [pytorch/pytorch#37030]
Currently, DDP creates buckets to consolidate gradient communications. This helps to reduce the total communication delay, but increases the memory footprint. We can potentially let DDP set `param.grad` to point to different offsets in the bucket, so that `AccumulateGrad` can update grads in place. In this way, we can avoid two copies of grads and at the same time avoid conducting copy operations to and from the bucket.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @xush6528 @osalpekar

**Comment by zhaojuanmao** [comment]
@mrshenli I'm wondering how to make param.grad to point to offsets in bucket and meanwhile send out the data to all reduce? are you saying to let bucket content and param.grad to share the same storage? 

I saw comments in codes to avoid that, see 

"    // Assert that the grad tensor and the bucket don't share storage.
    // If they did, we could avoid the copy altogether.
    // The reason for not doing this is that existing code calls
    // `detach_` from `zero_grad`, which is incompatible with views."

But I do not find where 'zero_grad' is called. do you have context about it?

**Comment by mrshenli** [comment]
> But I do not find where 'zero_grad' is called. do you have context about it?

`optimizer.zero_grad()` is usually called before calling `optimizer.step()` to update parameters.

**Comment by mrshenli** [comment]
> `detach_` from `zero_grad`, which is incompatible with views

I see, this is a valid reason to prevent us from using views as grad.

@albanD @ezyang Are my following understanding correct?

1. As views do not own storage and is just a way to interpret storage, it needs the edge to find the original tensor that owns the storage and hence cannot be detached in place from the graph?
2. We are doing `detach_()` instead of `detach()` in `optimizer.zero_grad()` to avoid the overhead of creating a new Tensor ptr?

Do you see any potential solution to get around this?

**Comment by ezyang** [comment]
I don't think it has to do with overhead, I suspect it has to do with ensuring old reference to the parameter are retained. But I'm not 100% sure, why don't you try the suggested change and see what happens?

**Comment by albanD** [comment]
1. The problem is that the view/inplace logic links the history of all these Tensors together. And so doing a `.detach_()` would be tricky. For example we allow it for the base even if it has views and this lead to the following weird behavior:
```python
import torch

a = torch.rand(10, requires_grad=True).clone()
b = a.narrow(0, 0, 2)

print("a: ", a.requires_grad, a.grad_fn)
print("b: ", b.requires_grad, b.grad_fn)

print("Detaching a inplace")
a.detach_()

print("a: ", a.requires_grad, a.grad_fn)
print("b: ", b.requires_grad, b.grad_fn)

print("Modifying b inplace")
b += 1

print("a: ", a.requires_grad, a.grad_fn)
print("b: ", b.requires_grad, b.grad_fn)
```
Giving
```
a:  True <CloneBackward object at 0x100e18350>
b:  True <SliceBackward object at 0x100e18350>
Detaching a inplace
a:  False None
b:  True <SliceBackward object at 0x100e18350>
Modifying b inplace
a:  True <CopySlices object at 0x100e18390>
b:  True <AsStridedBackward object at 0x100e18390>
```

2. We do detach_ to try and keep references to the old Tensor valid yes. And it is also slightly more efficient if you don't need a new Tensor object to do this as you only set 2 fields.

**Comment by mrshenli** [comment]
@ezyang thanks, we will try

@albanD the 1st bullet indeed looks weird. IIUC, modifying `b` inplace leads to `a` copied the first 2 elements from the original tensor and created a new tensor, and also changed the view (`b`) to point to the new tensor? Is that where the `CopySlices` come from? After doing this, is the original 10-element tensor still alive? BTW, is this a by-design behavior, or should we instead throw an error when trying to modify b inplace while a is detached?

**Comment by pritamdamania87** [comment]
Since we call `.detach()` on the grad in AccumulateGrad, what is the reason to detach it again in `zero_grad`? Is this for the case where the user might mess around with the .grad field (by setting requires_grad to True and doing a few ops) before calling zero_grad?

**Comment by albanD** [comment]
We only detach() in AccumulateGrad if there is no existing .grad and `create_graph=False`. So not always.

I think the main reason to detach() when you call `zero_grad` is to help the user avoid leaking memory. When they work with higher order derivatives, this make sure that we break the reference cycle created here (we have an issue to remove it, but no solution). Without this detach(), we would leak a lot of memory because the garbage collector wouldn't be able to break that link. See #3818 for the memory leak.
Beyond that, this is also doing a small optimization of deleting the graph that created the current .grad because the .zero_() overrides all the values anyway and so all these gradients will be 0.

**Comment by zhaojuanmao** [comment]
Hey @albanD, I tried to let grad point to DDP bucket buffers, in this case, variable.grad() will be view/slice of bucket buffers.  

I tried to call optimizer.zero_grad() after that, it failed because view can not call detach_(). But I tried to call detach() in optimizer.zero_grad(), it worked fine. 

I'm wondering whether we can change optimizer.zero_grad() code, check if param.grad is a view, call detach(), otherwise call detach_()?  if we call detach() for param.grad view, what is the impact here? 

Thanks!

**Comment by albanD** [comment]
Hi,

`.detach()` is out of place. So it is the same as doing nothing if you don't use the returned Tensor.
In this case I don't think we can go without the inplace detach.
That being said, you might be able to get a nicer behavior by doing:
If it does not have a grad_fn, do nothing.
If it does have a grad_fn,
  If it is a view: raise a nice error saying that your param's grad cannot be a view that requires grad
  If it is not a view: do the inplace detach

This should allow this to work in all the the use cases you care about right? Or do you explicitly need to support `create_graph=True`?

**Comment by jcyk** [comment]
Hi @zhaojuanmao @mrshenli ,
I would like to bring your attention to the following use case. 
Say I want to use DDP only for inference (no backward, no grads). Currently, DDP consumes much more additional and unnecessary memory (per "Currently, DDP creates buckets to consolidate gradient communications."). Can we pass an argument for DDP to disable this behavior?

Or is there any other workaround?

I noted in my experiments that the additional memory consumed is very huge when the model is huge!

**Comment by zhaojuanmao** [comment]
@jcyk if you do not need to sync grads, I guess you do not need to use DDP at all? you can just distribute your data and run inference locally

**Comment by zhaojuanmao** [comment]
closed by #44344

**Comment by jcyk** [comment]
@zhaojuanmao You are right. 
but then I cannot make use of the convenience of DDP for distributed computing and the code will be complicated. 

I think this is important for model distillation, where a (often big) teacher model only need to be do inference and a student model will learn to predict the teacher‘s output. 

In the above scenario, i need DDP for training but I don’t want to waste memory to hold unnecessary things for the teacher model.
